habitat_sim/simulator.py

```diff
@@ -244,43 +288,124 @@ def initialize_agent(
+
+        # As backport. All Dicts are ordered in Python >= 3.7
+        observations: Dict[int, Dict[str, Union[ndarray, "Tensor"]]] = OrderedDict()
+        for agent_id in agent_ids:
+            agent_observations: Dict[str, Union[ndarray, "Tensor"]] = {}
+            for sensor_uuid, sensor in self.__sensors[agent_id].items():
+                agent_observations[sensor_uuid] = sensor.get_observation()
+            observations[agent_id] = agent_observations
+        if return_single:
+            return next(iter(observations.values()))
+        return observations
 
-        observations = {}
-        for sensor_uuid, sensor in self._sensors.items():
-            observations[sensor_uuid] = sensor.get_observation()
+    @property
+    def _default_agent(self) -> Agent:
+        # TODO Deprecate and remove
+        return self.get_agent(agent_id=self._default_agent_id)
 
-        return observations
+    @property
+    def _last_state(self) -> AgentState:
+        # TODO Deprecate and remove
+        return self.__last_state[self._default_agent_id]
+
+    @_last_state.setter
+    def _last_state(self, state: AgentState) -> None:
+        # TODO Deprecate and remove
+        self.__last_state[self._default_agent_id] = state
 
-    def last_state(self):
-        return self._last_state
+    @property
+    def _sensors(self) -> Dict[str, "Sensor"]:
+        # TODO Deprecate and remove
+        return self.__sensors[self._default_agent_id]
 
+    def last_state(self, agent_id: Optional[int] = None) -> AgentState:
+        if agent_id is None:
+            agent_id = self._default_agent_id
+        return self.__last_state[agent_id]
+
+    @overload
     def step(
-        self, action: Any, dt: float = 1.0 / 60.0
+        self, action: Union[str, int], dt: float = 1.0 / 60.0
     ) -> Dict[str, Union[bool, ndarray, "Tensor"]]:
+        ...
+
+    @overload
+    def step(
+        self, action: MutableMapping_T[int, Union[str, int]], dt: float = 1.0 / 60.0
+    ) -> Dict[int, Dict[str, Union[bool, ndarray, "Tensor"]]]:
+        ...
+
+    def step(
+        self,
+        action: Union[str, int, MutableMapping_T[int, Union[str, int]]],
+        dt: float = 1.0 / 60.0,
+    ) -> Union[
+        Dict[str, Union[bool, ndarray, "Tensor"]],
+        Dict[int, Dict[str, Union[bool, ndarray, "Tensor"]]],
+    ]:
         self._num_total_frames += 1
-        collided = self._default_agent.act(action)
-        self._last_state = self._default_agent.get_state()
+        if isinstance(action, MutableMapping):
+            return_single = False
+        else:
+            action = cast(Dict[int, Union[str, int]], {self._default_agent_id: action})
+            return_single = True
+        collided_dict: Dict[int, bool] = {}
+        for agent_id, agent_act in action.items():
+            agent = self.get_agent(agent_id)
+            collided_dict[agent_id] = agent.act(agent_act)
+            self.__last_state[agent_id] = agent.get_state()
```

Comment:
Worth to make `__last_state` a dict as well to avoid handling list vs dict issues, like out of index.

Reply:
get_agent will also currently cause out of index, so I don't see the problem?

Reply:
If we really wanted it to be clean, we would move last state handling into the agent itself. There is no reason for the Sim to keep track of it. The agent can perfectly keep track of it.